pkg/generic/binarypb_codec.go

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2024 CloudWeGo Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package generic
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/cloudwego/kitex/pkg/remote"
+	"github.com/cloudwego/kitex/pkg/remote/codec"
+	"github.com/cloudwego/kitex/pkg/remote/codec/perrors"
+	"github.com/cloudwego/kitex/pkg/serviceinfo"
+)
+
+const (
+	metaInfoFixLen = 8
+)
+
+var _ remote.PayloadCodec = &binaryProtobufCodec{}
+
+type binaryPbReqType = []byte
+
+type binaryProtobufCodec struct {
+	protobufCodec remote.PayloadCodec
+}
+
+func (c *binaryProtobufCodec) Marshal(ctx context.Context, msg remote.Message, out remote.ByteBuffer) error {
+	data := msg.Data()
+	if data == nil {
+		return perrors.NewProtocolErrorWithMsg("invalid marshal data in rawProtobufBinaryCodec: nil")
+	}
+	if msg.MessageType() == remote.Exception {
+		if err := c.protobufCodec.Marshal(ctx, msg, out); err != nil {
+			return perrors.NewProtocolErrorWithMsg(fmt.Sprintf("rawProtobufBinaryCodec Marshal exception failed, err: %s", err.Error()))
+		}
+		return nil
+	}
+
+	methodName := msg.RPCInfo().Invocation().MethodName()
+	if methodName == "" {
+		return errors.New("empty methodName in protobuf Marshal")
+	}
+	// 3. encode metainfo
+	// 3.1 magic && msgType
+	if err := codec.WriteUint32(codec.ProtobufV1Magic+uint32(msg.MessageType()), out); err != nil {
+		return perrors.NewProtocolErrorWithMsg(fmt.Sprintf("protobuf marshal, write meta info failed: %s", err.Error()))
+	}
+	// 3.2 methodName
+	if _, err := codec.WriteString(methodName, out); err != nil {
+		return perrors.NewProtocolErrorWithMsg(fmt.Sprintf("protobuf marshal, write method name failed: %s", err.Error()))
+	}
+	// 3.3 seqID
+	if err := codec.WriteUint32(uint32(msg.RPCInfo().Invocation().SeqID()), out); err != nil {
+		return perrors.NewProtocolErrorWithMsg(fmt.Sprintf("protobuf marshal, write seqID failed: %s", err.Error()))
+	}
```

Comment:
This Marshal method is only for payload (kitex writes metainfo before encoding), so you don't need to encode metainfo.
But when it's client, binary generic needs to set seqID because the users do encode to create a request, but they don't care about seqID.
You can refer to binary thrift generic codec.

Reply:
And you also need to prepare request including metainfo for tests.